ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java

```diff
@@ -0,0 +1,1240 @@
+    if (node.getLocation().isPresent()) {
+      return new InPredicate(node.getLocation().get(),
+          (Expression) process(node.getValue(), context),
+          (Expression) process(node.getValueList(), context));
+    } else {
+      return new InPredicate((Expression) process(node.getValue(), context),
+          (Expression) process(node.getValueList(), context));
+    }
+
+  }
+
+  protected Node visitFunctionCall(FunctionCall node, Object context) {
+    Optional<Window> window = node.getWindow().isPresent()
+        ? Optional.ofNullable((Window) process(node.getWindow().get(), context))
+        : Optional.empty();
+
+    if (node.getLocation().isPresent()) {
+      return new FunctionCall(node.getLocation().get(),
+          node.getName(),
+          window,
+          node.isDistinct(),
+          node.getArguments()
+              .stream()
+              .map(arg -> (Expression) process(arg, context))
+              .collect(Collectors.toList())
+      );
+    } else {
+      return new FunctionCall(node.getName(),
+          window,
+          node.isDistinct(),
+          node.getArguments()
+              .stream()
+              .map(arg -> (Expression) process(arg, context))
+              .collect(Collectors.toList())
+      );
+    }
+
+  }
+
+  protected Node visitLambdaExpression(LambdaExpression node, Object context) {
+    if (node.getLocation().isPresent()) {
+      return new LambdaExpression(node.getLocation().get(),
+          node.getArguments(),
+          (Expression) process(node.getBody(), context)
+      );
+    } else {
+      return new LambdaExpression(node.getArguments(),
+          (Expression) process(node.getBody(), context)
+      );
+    }
+  }
+
+  protected Node visitSimpleCaseExpression(SimpleCaseExpression node, Object context) {
+    Optional<Expression> defaultValue = node.getDefaultValue().isPresent()
+        ? Optional.ofNullable((Expression) process(node.getDefaultValue().get(), context))
+        : Optional.empty();
+    if (node.getLocation().isPresent()) {
+      return new SimpleCaseExpression(node.getLocation().get(),
+          (Expression) process(node.getOperand(), context),
+          node.getWhenClauses()
+              .stream()
+              .map(whenClause ->
+                  (WhenClause) process(whenClause, context))
+              .collect(Collectors.toList()),
+          defaultValue
+      );
+    } else {
+      return new SimpleCaseExpression((Expression) process(node.getOperand(), context),
+          node.getWhenClauses()
+              .stream()
+              .map(whenClause ->
+                  (WhenClause) process(whenClause, context))
+              .collect(Collectors.toList()),
+          defaultValue
+      );
+    }
+
+  }
+
+  protected Node visitStringLiteral(StringLiteral node, Object context) {
+    if (node.getLocation().isPresent()) {
+      return new StringLiteral(node.getLocation().get(),
```

Comment:
If we're not processing any sub-nodes why create a new node here? Ditto for other visit functions that are just copying the node.

Reply:
Rewrite will create a new AST.

Reply:
Thats what I was thinking. It just seemed odd to me that for some nodes we'll copy the node and for others we'd reuse the old one. I don't think it makes a difference from a correctness pov though.